librad/src/git/types.rs

```diff
@@ -0,0 +1,269 @@
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use std::{
+    collections::{HashMap, HashSet},
+    fmt::{self, Display},
+};
+
+use crate::{git::refs::Refs, hash::Hash, peer::PeerId, uri::RadUrn};
+
+pub type Namespace = Hash;
+
+#[derive(Clone, Copy)]
+pub enum RefsCategory {
+    Heads,
+    Rad,
+}
+
+impl Display for RefsCategory {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Self::Heads => f.write_str("heads"),
+            Self::Rad => f.write_str("rad"),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct Reference {
+    pub namespace: Namespace,
+    pub remote: Option<PeerId>,
+    pub category: RefsCategory,
+    pub name: String, // TODO: apply validation like `uri::Path`
+}
+
+impl Display for Reference {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "refs/namespaces/{}/refs/", self.namespace)?;
+
+        match &self.remote {
+            None => write!(f, "{}/{}", self.category, self.name),
+            Some(remote) => write!(f, "remotes/{}/{}/{}", remote, self.category, self.name),
+        }
+    }
+}
+
+impl Reference {
+    pub fn rad_id(namespace: Namespace) -> Self {
+        Self {
+            namespace,
+            remote: None,
+            category: RefsCategory::Rad,
+            name: "id".to_owned(),
+        }
+    }
+
+    pub fn rad_ids_glob(namespace: Namespace) -> Self {
+        Self {
+            namespace,
+            remote: None,
+            category: RefsCategory::Rad,
+            name: "ids/*".to_owned(),
+        }
+    }
+
+    pub fn rad_refs(namespace: Namespace, remote: impl Into<Option<PeerId>>) -> Self {
+        Self {
+            namespace,
+            remote: remote.into(),
+            category: RefsCategory::Rad,
+            name: "refs".to_owned(),
+        }
+    }
+
+    pub fn head(namespace: Namespace, remote: impl Into<Option<PeerId>>, name: &str) -> Self {
+        Self {
+            namespace,
+            remote: remote.into(),
+            category: RefsCategory::Heads,
+            name: name.to_owned(),
+        }
+    }
+
+    pub fn set_remote(mut self, remote: impl Into<Option<PeerId>>) -> Self {
+        self.remote = remote.into();
+        self
+    }
+
+    pub fn with_remote(&self, remote: impl Into<Option<PeerId>>) -> Self {
+        Self {
+            remote: remote.into(),
+            ..self.clone()
+        }
+    }
+
+    pub fn set_name(mut self, name: &str) -> Self {
+        self.name = name.to_owned();
+        self
+    }
+
+    pub fn with_name(&self, name: &str) -> Self {
+        Self {
+            name: name.to_owned(),
+            ..self.clone()
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct Refspec {
+    pub remote: Reference,
+    pub local: Reference,
+    pub force: bool,
```

Comment:
I think I'd prefer an `enum` instead of a `bool`. What do you think?

Reply:
You have a strange aversion against Mr. Boole 😛 Why would an `enum` be preferable?

Reply:
Because I know I'll come along and get it the wrong way around one day :sweat_smile: Whereas if I have a type to work with I won't (less likely) get them the wrong way around. I've got bad boolean sight :grin:

Reply:
But you'd always construct it using

```rust
Refspec {
    remote,
    local,
    force: jawohl
}
```

so I'm not sure

Reply:
Ya, and I suppose it's only used in the `Display` impl anyway. No biggy.